core-rust/node-common/src/locks.rs

```diff
@@ -235,6 +243,103 @@ impl<'a, T> Drop for RwLockWriteGuard<'a, T> {
     }
 }
 
+/// A custom lock primitive guarding a "current state" of a value composed from an (immutable)
+/// "historical" and (live) "current" parts of state.
+/// The assumption is that the current state needs a classic [`RwLock`] access, while the historical
+/// state can be accessed freely, without obtaining any lock.
+/// The lock caller is responsible for distinguishing proper current vs historical access.
+// TODO(future refactoring): It seems like the "weird lock" should not be needed if we had a DB
+// interface which is more aware of its "current vs historical" nature. Maybe we will naturally go
+// in that direction when introducing DB snapshotting.
+pub struct StateLock<T> {
+    underlying: RwLock<()>, // we use our own primitive lock a marker for current state
+    value: UnsafeCell<T>,   // we need unsafe access across threads
+}
+
+unsafe impl<T: Send> Send for StateLock<T> {}
+unsafe impl<T: Send + Sync> Sync for StateLock<T> {}
+
+impl<T> StateLock<T> {
+    /// Locks the current state for reading.
+    /// This method should be used when caller needs a series of reads referring to the current
+    /// state (it "freezes" the notion of "current").
+    pub fn read_current(&self) -> StateLockReadGuard<'_, T> {
+        StateLockReadGuard {
+            underlying: self.underlying.read(),
+            value: &self.value,
+        }
+    }
+
+    /// Locks the current state for writing.
+    /// This method should be used when caller wants to update the guarded value in a way which
+    /// changes the notion of "current".
+    pub fn write_current(&self) -> StateLockWriteGuard<'_, T> {
+        StateLockWriteGuard {
+            underlying: self.underlying.write(),
+            value: &self.value,
+        }
+    }
+
+    /// Returns a direct reference to the guarded value, without locking anything.
+    /// This method should be used when caller wants to interact selectively with pieces of
+    /// historical state.
+    pub fn access_historical(&self) -> &T {
+        unsafe { &*self.value.get() }
```

Comment:
Soooo this is technically UB. We allow access to &T and &mut T at the same time.

I don't think this is a problem... But do we actually need to allow access to &mut T for writes?

Or can we just store a T instead of an `UnsafeCell<T>` and just return `&T` in either case, and leave it up to the client to choose which they need?

Reply:
As discussed on our call - I went the "get rid of memoryDb" way (since it was the only thing requiring the `&mut`). The change has grown considerably, PTAL.

_(I ack that some cleanups are now possible after that removal [and I even left one TODO behind], but I don't want to propagate changes even further)_